Support for JS translations in extensions

torfsen opened the pull request:
This PR is a WIP to support JavaScript translations in extensions (ckan/ideas-and-roadmap#176).

Currently, CKAN's translation files for JavaScript are generated "manually" using `paster trans js`. The generated files take only the translations from core CKAN into account.

This PR introduces the following changes:
- Translations from extensions that implement `ITranslation` are taken into account. Hence, extensions may override CKAN translations used in JS or introduce their own translatable strings in JS.
- Translations are only (re-)generated if they're out-of-date, i.e. if there's a PO-file for the language that's newer than the generated JS file.
- Translations are re-generated during server startup (if necessary).

With these changes, translations in extensions work transparently for both Jinja and JS. No changes to the extensions are necessary. Running `paster trans js` isn't required anymore, but the command has been kept in case one wants to work on translations without constantly restarting the server.

I'd love to get some feedback on that approach. Once there's a consensus I'll add the missing tests and update the documentation.

torfsen commented:
Note: The failing test is due to #3270 which will be fixed by #3246.

torfsen commented:
@TkTech Some feedback from you on this would be much appreciated.

amercader commented:
@torfsen this looks great. Happy to review it fully and merge it if you sort tests and docs. Cheers

torfsen commented:
@amercader I've added tests and updated the documentation so this is ready to go from my point of view.

I didn't find an established convention of how to bundle extra files with test cases, so I created the subdirectory `ckan/tests/lib/_i18n_build_js_translations` for them. Let me know if there's a better place.

amercader commented:
This all looks great @torfsen. There is already an example `ITranslation` plugin with tests on `ckanext/example_itranslation`. Any reason why we can add the new tests there?

torfsen commented:
@amercader Only some of the tests target `ITranslation`, so I'd rather keep most of them where they currently are. I could re-use the `example_itranslation` plugin for the `ITranslation`-related tests, but then I would have to include the necessary edge case translation strings in its `.pot` and `.po` files, where they might confuse people who just want to learn from the example.

Hence, in my opinion it's cleaner to keep the tests as they are, but at the end it's probably a matter of taste. Let me know if you'd rather have them as part of `example_itranslation`.

amercader commented:
Thanks @torfsen !
